File: src/bin/liq_kamino_monitor.rs
//! Kamino continuous liquidation-opportunity monitor (read-only, long-run).
//!
//! Full-scans obligations occasionally to build a watch-set of near-liquidation
//! positions, then polls only that set + its reserves frequently — recomputing
//! health at each reserve's latest cached price. Logs when a position crosses
//! liquidatable (APPEARED) and when it's taken/recovers (RESOLVED, with survival
//! seconds). Over hours this measures real opportunity flow + how fast the
//! competition takes them = the go/no-go for a Kamino liquidation executor.
//!
//! Usage: HELIUS_RPC=<url> [POLL_SECS=15] [FULL_SCAN_SECS=300] [WATCH_RATIO=0.92]
//!        [MIN_COLLATERAL_USD=100] cargo run --release --bin liq_kamino_monitor

use arb_engine::kamino::{self, Obligation, Reserve};
use solana_pubkey::Pubkey;
use std::collections::{HashMap, HashSet};
use std::io::Write;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

fn now_ts() -> u64 { SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs() }

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(resp) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = resp.into_json::<serde_json::Value>() { return Some(v); }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}
fn b64(d: &serde_json::Value) -> Option<Vec<u8>> {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.decode(d.get(0)?.as_str()?).ok()
}
fn get_multiple(endpoint: &str, keys: &[Pubkey], slice_len: u64) -> HashMap<Pubkey, Vec<u8>> {
    let mut out = HashMap::new();
    for chunk in keys.chunks(100) {
        let strs: Vec<String> = chunk.iter().map(|k| k.to_string()).collect();
        let Some(v) = rpc(endpoint, serde_json::json!({
            "jsonrpc":"2.0","id":1,"method":"getMultipleAccounts",
            "params":[strs, {"encoding":"base64","dataSlice":{"offset":0,"length":slice_len}}]
        })) else { continue };
        for (i, acc) in v["result"]["value"].as_array().into_iter().flatten().enumerate() {
            if let Some(bytes) = acc.get("data").and_then(b64) { out.insert(chunk[i], bytes); }
        }
        std::thread::sleep(Duration::from_millis(40));
    }
    out
}

/// Fetch reserves referenced by a set of obligations, decode → map.
fn fetch_reserves(endpoint: &str, obs: &[(Pubkey, Obligation)]) -> HashMap<Pubkey, Reserve> {
    let pks: Vec<Pubkey> = obs.iter()
        .flat_map(|(_, o)| o.deposits.iter().map(|d| d.0).chain(o.borrows.iter().map(|b| b.0)))
        .collect::<HashSet<_>>().into_iter().collect();
    get_multiple(endpoint, &pks, 5016).iter()
        .filter_map(|(pk, raw)| Reserve::decode(raw).map(|r| (*pk, r)))
        .collect()
}

/// Full scan → all debt obligations (with their account pubkey).
fn full_scan(endpoint: &str) -> Vec<(Pubkey, Obligation)> {
    let resp = rpc(endpoint, serde_json::json!({
        "jsonrpc":"2.0","id":1,"method":"getProgramAccounts",
        "params":[kamino::KLEND_PROGRAM, {"encoding":"base64","dataSlice":{"offset":0,"length":2288},
            "filters":[{"dataSize": kamino::OBLIGATION_SIZE}]}]
    }));
    resp.as_ref().and_then(|v| v["result"].as_array()).cloned().unwrap_or_default().iter()
        .filter_map(|e| {
            let pk = e["pubkey"].as_str()?.parse::<Pubkey>().ok()?;
            let o = Obligation::decode(&b64(&e["account"]["data"])?)?;
            if o.borrows.is_empty() { None } else { Some((pk, o)) }
        }).collect()
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC").or_else(|_| std::env::var("RPC_HTTP")).expect("HELIUS_RPC");
    let poll = Duration::from_secs(std::env::var("POLL_SECS").ok().and_then(|s| s.parse().ok()).unwrap_or(15));
    let full_scan_secs: u64 = std::env::var("FULL_SCAN_SECS").ok().and_then(|s| s.parse().ok()).unwrap_or(300);
    let watch_ratio: f64 = std::env::var("WATCH_RATIO").ok().and_then(|s| s.parse().ok()).unwrap_or(0.92);
    let min_collateral: f64 = std::env::var("MIN_COLLATERAL_USD").ok().and_then(|s| s.parse().ok()).unwrap_or(100.0);

    let _ = std::fs::create_dir_all("runs");
    let mut log = std::fs::OpenOptions::new().create(true).append(true).open("runs/kamino_opportunities.jsonl")
        .expect("open log");
    macro_rules! emit { ($v:expr) => {{ let _ = writeln!(log, "{}", $v); let _ = log.flush(); }} }

    eprintln!("[kmon] poll={}s full_scan={}s watch_ratio={} min_collateral=${}",
        poll.as_secs(), full_scan_secs, watch_ratio, min_collateral);

    // watch-set: account pubkey → obligation (positions refreshed each poll)
    let mut watch: HashMap<Pubkey, Obligation> = HashMap::new();
    let mut last_full = 0u64;
    let mut open: HashMap<Pubkey, (u64, f64, String)> = HashMap::new(); // acct → (first_ts, peak_collateral, owner)
    let (mut appeared, mut resolved) = (0u64, 0u64);
    let mut durations: Vec<u64> = Vec::new();

    loop {
        // Periodic full scan → refresh watch-set (near-liquidation + trustworthy).
        if now_ts().saturating_sub(last_full) >= full_scan_secs {
            eprintln!("[kmon {}] full scan …", now_ts());
            let all = full_scan(&endpoint);
            let reserves = fetch_reserves(&endpoint, &all);
            let mut new_watch = HashMap::new();
            for (pk, o) in all {
                let r = kamino::recompute(&o, &reserves);
                if r.trustworthy() && r.deposited_value >= min_collateral && r.ratio() >= watch_ratio {
                    new_watch.insert(pk, o);
                }
            }
            eprintln!("[kmon {}] watch-set: {} obligations (ratio ≥ {}, collateral ≥ ${})",
                now_ts(), new_watch.len(), watch_ratio, min_collateral);
            watch = new_watch;
            last_full = now_ts();
        }

        // Poll: fresh obligation data + fresh reserve prices → recompute.
        let watch_pks: Vec<Pubkey> = watch.keys().copied().collect();
        for (pk, raw) in get_multiple(&endpoint, &watch_pks, 2288) {
            if let Some(o) = Obligation::decode(&raw) { watch.insert(pk, o); }
        }
        let obs: Vec<(Pubkey, Obligation)> = watch.iter().map(|(k, v)| (*k, v.clone())).collect();
        let reserves = fetch_reserves(&endpoint, &obs);

        let ts = now_ts();
        let mut cur_liq = 0usize;
        let mut tightest = 0.0f64;
        for (pk, o) in &obs {
            let r = kamino::recompute(o, &reserves);
            if !r.trustworthy() { continue; }
            tightest = tightest.max(r.ratio());
            if r.liquidatable() && r.deposited_value >= min_collateral {
                cur_liq += 1;
                let e = open.entry(*pk).or_insert_with(|| {
                    appeared += 1;
                    emit!(serde_json::json!({"ts":ts,"event":"appeared","protocol":"kamino",
                        "account":pk.to_string(),"owner":o.owner.to_string(),
                        "collateral_usd":(r.deposited_value*100.0).round()/100.0,
                        "debt_usd":(r.bf_adjusted_debt*100.0).round()/100.0,
                        "ratio":(r.ratio()*10000.0).round()/10000.0}));
                    eprintln!("[APPEARED {}] {}… collateral=${:.0} ratio={:.4}", ts, &o.owner.to_string()[..8], r.deposited_value, r.ratio());
                    (ts, r.deposited_value, o.owner.to_string())
                });
                e.1 = e.1.max(r.deposited_value);
            } else if let Some((t0, peak, owner)) = open.remove(pk) {
                resolved += 1;
                let dur = ts.saturating_sub(t0);
                durations.push(dur);
                emit!(serde_json::json!({"ts":ts,"event":"resolved","protocol":"kamino",
                    "account":pk.to_string(),"owner":owner,"seen_secs":dur,
                    "peak_collateral_usd":(peak*100.0).round()/100.0}));
                eprintln!("[RESOLVED {}] {}… after {}s (peak ${:.0})", ts, &owner[..8], dur, peak);
            }
        }
        durations.sort_unstable();
        let med = durations.get(durations.len()/2).copied().unwrap_or(0);
        eprintln!("[kmon {ts}] watch={} liq_now={} open={} | appeared={} resolved={} med_survival={}s | tightest {:.4}",
            watch.len(), cur_liq, open.len(), appeared, resolved, med, tightest);

        std::thread::sleep(poll);
    }
}

File: src/bin/liqKaminoMonitor.ts
// Port of src/bin/liq_kamino_monitor.rs
//
// Kamino continuous liquidation-opportunity monitor (read-only, long-run).
//
// Full-scans obligations occasionally to build a watch-set of near-liquidation
// positions, then polls only that set + its reserves frequently — recomputing
// health at each reserve's latest cached price. Logs when a position crosses
// liquidatable (APPEARED) and when it's taken/recovers (RESOLVED, with survival
// seconds). Over hours this measures real opportunity flow + how fast the
// competition takes them = the go/no-go for a Kamino liquidation executor.
//
// Usage: HELIUS_RPC=<url> [POLL_SECS=15] [FULL_SCAN_SECS=300] [WATCH_RATIO=0.92]
//        [MIN_COLLATERAL_USD=100] tsx src/bin/liqKaminoMonitor.ts

import 'dotenv/config';
import * as fs from 'node:fs';
import { PublicKey } from '@solana/web3.js';
import {
  decodeObligation,
  decodeReserve,
  KLEND_PROGRAM,
  OBLIGATION_SIZE,
  recompute,
  recomputedLiquidatable,
  recomputedRatio,
  recomputedTrustworthy,
  type Obligation,
  type ReserveMap,
} from '../lib/kamino.js';

function nowTs(): number {
  return Math.floor(Date.now() / 1000);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await res.json();
    } catch {
      // fall through to retry
    }
    await sleep(400 << attempt);
  }
  return undefined;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function b64(data: any): Buffer | undefined {
  const s = data?.[0];
  if (typeof s !== 'string') return undefined;
  try {
    return Buffer.from(s, 'base64');
  } catch {
    return undefined;
  }
}

async function getMultiple(endpoint: string, keys: PublicKey[], sliceLen: number): Promise<Map<string, Buffer>> {
  const out = new Map<string, Buffer>();
  for (let i = 0; i < keys.length; i += 100) {
    const chunk = keys.slice(i, i + 100);
    const strs = chunk.map((k) => k.toBase58());
    const v = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'getMultipleAccounts',
      params: [strs, { encoding: 'base64', dataSlice: { offset: 0, length: sliceLen } }],
    });
    const arr: unknown[] = Array.isArray(v?.result?.value) ? v.result.value : [];
    arr.forEach((acc, j) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const bytes = (acc as any)?.data !== undefined ? b64((acc as any).data) : undefined;
      if (bytes !== undefined) out.set(chunk[j].toBase58(), bytes);
    });
    await sleep(40);
  }
  return out;
}

/** Fetch reserves referenced by a set of obligations, decode -> map. */
async function fetchReserves(endpoint: string, obs: Array<[PublicKey, Obligation]>): Promise<ReserveMap> {
  const pkSet = new Set<string>();
  const pks: PublicKey[] = [];
  for (const [, o] of obs) {
    for (const [r] of o.deposits) {
      if (!pkSet.has(r.toBase58())) {
        pkSet.add(r.toBase58());
        pks.push(r);
      }
    }
    for (const [r] of o.borrows) {
      if (!pkSet.has(r.toBase58())) {
        pkSet.add(r.toBase58());
        pks.push(r);
      }
    }
  }
  const raw = await getMultiple(endpoint, pks, 5016);
  const out: ReserveMap = new Map();
  for (const [pk, bytes] of raw) {
    const r = decodeReserve(bytes);
    if (r !== null) out.set(pk, r);
  }
  return out;
}

/** Full scan -> all debt obligations (with their account pubkey). */
async function fullScan(endpoint: string): Promise<Array<[PublicKey, Obligation]>> {
  const resp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getProgramAccounts',
    params: [
      KLEND_PROGRAM,
      { encoding: 'base64', dataSlice: { offset: 0, length: 2288 }, filters: [{ dataSize: OBLIGATION_SIZE }] },
    ],
  });
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const entries: any[] = Array.isArray(resp?.result) ? resp.result : [];
  const out: Array<[PublicKey, Obligation]> = [];
  for (const e of entries) {
    let pk: PublicKey;
    try {
      pk = new PublicKey(e?.pubkey);
    } catch {
      continue;
    }
    const bytes = e?.account?.data !== undefined ? b64(e.account.data) : undefined;
    if (bytes === undefined) continue;
    const o = decodeObligation(bytes);
    if (o !== null && o.borrows.length > 0) out.push([pk, o]);
  }
  return out;
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP;
  if (!endpoint) throw new Error('HELIUS_RPC');
  const pollSecs = Number.parseInt(process.env.POLL_SECS ?? '', 10) || 15;
  const fullScanSecs = Number.parseInt(process.env.FULL_SCAN_SECS ?? '', 10) || 300;
  const watchRatio = Number.parseFloat(process.env.WATCH_RATIO ?? '') || 0.92;
  const minCollateral = Number.parseFloat(process.env.MIN_COLLATERAL_USD ?? '') || 100.0;

  fs.mkdirSync('runs', { recursive: true });
  const logFd = fs.openSync('runs/kamino_opportunities.jsonl', 'a');
  const emit = (v: unknown): void => {
    fs.writeSync(logFd, `${JSON.stringify(v)}\n`);
  };

  console.error(`[kmon] poll=${pollSecs}s full_scan=${fullScanSecs}s watch_ratio=${watchRatio} min_collateral=$${minCollateral}`);

  // watch-set: account pubkey (base58) -> obligation (positions refreshed each poll)
  let watch = new Map<string, Obligation>();
  let lastFull = 0;
  // acct (base58) -> (first_ts, peak_collateral, owner)
  const open = new Map<string, { t0: number; peak: number; owner: string }>();
  let appeared = 0;
  let resolved = 0;
  const durations: number[] = [];

  for (;;) {
    // Periodic full scan -> refresh watch-set (near-liquidation + trustworthy).
    if (nowTs() - lastFull >= fullScanSecs) {
      console.error(`[kmon ${nowTs()}] full scan …`);
      const all = await fullScan(endpoint);
      const reserves = await fetchReserves(endpoint, all);
      const newWatch = new Map<string, Obligation>();
      for (const [pk, o] of all) {
        const r = recompute(o, reserves);
        if (recomputedTrustworthy(r) && r.depositedValue >= minCollateral && recomputedRatio(r) >= watchRatio) {
          newWatch.set(pk.toBase58(), o);
        }
      }
      console.error(`[kmon ${nowTs()}] watch-set: ${newWatch.size} obligations (ratio ≥ ${watchRatio}, collateral ≥ $${minCollateral})`);
      watch = newWatch;
      lastFull = nowTs();
    }

    // Poll: fresh obligation data + fresh reserve prices -> recompute.
    const watchPks = Array.from(watch.keys()).map((k) => new PublicKey(k));
    const raw = await getMultiple(endpoint, watchPks, 2288);
    for (const [pkStr, bytes] of raw) {
      const o = decodeObligation(bytes);
      if (o !== null) watch.set(pkStr, o);
    }
    const obs: Array<[PublicKey, Obligation]> = Array.from(watch.entries()).map(([k, v]) => [new PublicKey(k), v]);
    const reserves = await fetchReserves(endpoint, obs);

    const ts = nowTs();
    let curLiq = 0;
    let tightest = 0.0;
    for (const [pk, o] of obs) {
      const r = recompute(o, reserves);
      if (!recomputedTrustworthy(r)) continue;
      tightest = Math.max(tightest, recomputedRatio(r));
      const pkStr = pk.toBase58();
      if (recomputedLiquidatable(r) && r.depositedValue >= minCollateral) {
        curLiq += 1;
        let e = open.get(pkStr);
        if (e === undefined) {
          appeared += 1;
          emit({
            ts,
            event: 'appeared',
            protocol: 'kamino',
            account: pkStr,
            owner: o.owner.toBase58(),
            collateral_usd: Math.round(r.depositedValue * 100.0) / 100.0,
            debt_usd: Math.round(r.bfAdjustedDebt * 100.0) / 100.0,
            ratio: Math.round(recomputedRatio(r) * 10000.0) / 10000.0,
          });
          console.error(`[APPEARED ${ts}] ${o.owner.toBase58().slice(0, 8)}… collateral=$${r.depositedValue.toFixed(0)} ratio=${recomputedRatio(r).toFixed(4)}`);
          e = { t0: ts, peak: r.depositedValue, owner: o.owner.toBase58() };
          open.set(pkStr, e);
        }
        e.peak = Math.max(e.peak, r.depositedValue);
      } else {
        const e = open.get(pkStr);
        if (e !== undefined) {
          open.delete(pkStr);
          resolved += 1;
          const dur = Math.max(0, ts - e.t0);
          durations.push(dur);
          emit({
            ts,
            event: 'resolved',
            protocol: 'kamino',
            account: pkStr,
            owner: e.owner,
            seen_secs: dur,
            peak_collateral_usd: Math.round(e.peak * 100.0) / 100.0,
          });
          console.error(`[RESOLVED ${ts}] ${e.owner.slice(0, 8)}… after ${dur}s (peak $${e.peak.toFixed(0)})`);
        }
      }
    }
    durations.sort((a, b) => a - b);
    const med = durations.length > 0 ? durations[Math.floor(durations.length / 2)] : 0;
    console.error(
      `[kmon ${ts}] watch=${watch.size} liq_now=${curLiq} open=${open.size} | appeared=${appeared} resolved=${resolved} med_survival=${med}s | tightest ${tightest.toFixed(4)}`,
    );

    await sleep(pollSecs * 1000);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
